diff --git a/dependency_support/verilator/verilator.BUILD.bazel b/dependency_support/verilator/verilator.BUILD.bazel
index 7863081..474a126 100644
--- a/dependency_support/verilator/verilator.BUILD.bazel
+++ b/dependency_support/verilator/verilator.BUILD.bazel
@@ -405,6 +405,71 @@ cc_library(
     ],
 )
 
+
+cc_library(
+    name = "verilator_cpp",
+    srcs = [
+        "include/gtkwave/fastlz.h",
+        "include/gtkwave/fst_config.h",
+        "include/gtkwave/fstapi.h",
+        "include/gtkwave/lz4.h",
+        "include/gtkwave/wavealloca.h",
+        "include/verilated_fst_c.cpp",
+        "include/verilated_imp.h",
+        "include/verilated_syms.h",
+        "include/verilated_threads.cpp",
+        "include/verilated_vcd_c.cpp",
+        "include/verilated_vpi.cpp",
+        "include/verilated.cpp",
+    ],
+    hdrs = [
+        "include/verilated_config.h",
+        "include/verilated_dpi.h",
+        "include/verilated_fst_c.h",
+        "include/verilated_funcs.h",
+        "include/verilated_intrinsics.h",
+        "include/verilated_sym_props.h",
+        "include/verilated_threads.h",
+        "include/verilated_timing.h",
+        "include/verilated_trace_imp.h",
+        "include/verilated_trace.h",
+        "include/verilated_types.h",
+        "include/verilated_vcd_c.h",
+        "include/verilated_vpi.h",
+        "include/verilated.h",
+    ],
+    copts = [
+        "-Wno-unused-const-variable",
+        "-Wno-sign-compare",
+        "-Wno-unused-variable",
+        "-Wno-unused-parameter",
+        "-Wno-unused-but-set-variable",
+        "-std=c++17",
+        "-ffp-contract=off",
+    ],
+    includes = [
+        "_virtual_includes/verilator_cpp",
+        "_virtual_includes/vltstd",
+        "_virtual_includes/verilatedos",
+        "_virtual_includes/svdpi",
+    ],
+    linkopts = [
+        "-lpthread",
+    ],
+    strip_include_prefix = "include/",
+    textual_hdrs = [
+        "include/gtkwave/fastlz.c",
+        "include/gtkwave/fstapi.c",
+        "include/gtkwave/lz4.c",
+    ],
+    visibility = ["//visibility:public"],
+    deps = [
+        "@net_zlib//:zlib",
+        ":vltstd",
+        ":verilatedos",
+    ],
+)
+
 # This alias is for supporting the legacy name but while allowing the
 # `cc_library` target to be renamed to avoid a `liblibverilator.a`
 # output name.
diff --git a/verilator/BUILD.bazel b/verilator/BUILD.bazel
index e41b3d8..160b8ae 100644
--- a/verilator/BUILD.bazel
+++ b/verilator/BUILD.bazel
@@ -32,6 +32,11 @@ verilator_toolchain(
         "@verilator//:libverilator",
         "@verilator//:svdpi",
     ],
+    cpp_deps = [
+        "@net_zlib//:zlib",
+        "@verilator//:verilator_cpp",
+        "@verilator//:svdpi",
+    ],
 )
 
 toolchain(
diff --git a/verilator/defs.bzl b/verilator/defs.bzl
index c28ccf3..25dab41 100644
--- a/verilator/defs.bzl
+++ b/verilator/defs.bzl
@@ -182,7 +182,7 @@ def _verilator_cc_library(ctx):
         defines = defines,
         runfiles = runfiles,
         includes = [verilator_output_hpp.path],
-        deps = verilator_toolchain.deps,
+        deps = verilator_toolchain.deps if ctx.attr.systemc else verilator_toolchain.cpp_deps,
     )
 
 verilator_cc_library = rule(
@@ -241,6 +241,7 @@ def _verilator_toolchain_impl(ctx):
     return [platform_common.ToolchainInfo(
         verilator = ctx.executable.verilator,
         deps = ctx.attr.deps,
+        cpp_deps = ctx.attr.cpp_deps,
         extra_vopts = ctx.attr.extra_vopts,
         all_files = all_files,
     )]
@@ -253,6 +254,10 @@ verilator_toolchain = rule(
             doc = "Global Verilator dependencies to link into downstream targets.",
             providers = [CcInfo],
         ),
+        "cpp_deps": attr.label_list(
+            doc = "Verilator dependencies for non-SystemC C++ models.",
+            providers = [CcInfo],
+        ),
         "extra_vopts": attr.string_list(
             doc = "Extra flags to pass to Verilator compile actions.",
         ),
